### ISSUE TITLE
Possible Cause of BadDiscountException?

### ISSUE THREAD

**Issue by windweller** [kpu/kenlm#68]
```
.../libs/kenlm/lm/builder/adjust_counts.cc:61 in void lm::builder::{anonymous}::StatCollector::CalculateDiscounts(const lm::builder::DiscountConfig&) threw BadDiscountException because `discounts_[i].amount[j] < 0.0 || discounts_[i].amount[j] > j'.
ERROR: 1-gram discount out of range for adjusted count 2: -1.6000001
Aborted (core dumped)
```

What could have happened to cause this error? We preprocessed the files to limit to 10k vocab (replace out of vocab words with <unk>). The files are sufficiently big enough (with line breaks, thanks to the help in the other thread), some output info:

```
Unigram tokens 77187240 types 10002
=== 2/5 Calculating and sorting adjusted counts ===
Chain sizes: 1:120024 2:37614993408 3:70528114688
ERROR: 1-gram discount out of range for adjusted count 2: -1.75
Aborted (core dumped)
```

Did fixing the vocabulary externally cause this problem?

**Comment by kpu** [comment]
Modified Kneser-Ney smoothing is designed to model language, rare words and everything.  Specifically, it explicitly uses counts of singletons, doubletons, and tripletons and above to determine discounts (see Section 3.2 of https://kheafield.com/professional/edinburgh/estimate_paper.pdf ).  When you restrict the vocabulary and replace rare words, these singletons no longer exist, the formulae that depend on their counts break, and you get an error message.  

Unfortunately, much of the neural language modeling literature compares to Kneser-Ney smoothing in its handicapped setting without singletons (such as @kyunghyuncho in https://arxiv.org/pdf/1511.03729.pdf ).  With `lmplz`, you can do this by using `--discount_fallback`.  But I'd ask you to instead make an honest comparison of perplexity without handicapping the baseline.  

I suggest the task be to **model real language**.  Your test set has all the words in it with no clipping.  Run Kneser-Ney smoothing on the full, unmodified text.  If your neural network needs a smaller vocabulary, train that on clipped vocabulary.  That gives you a classifier over specific words and a class of clipped words.  Carry on by predicting the actual word given the class:
`p(clipped word class | neural network)p(actual word | clipped word class)`
This is the term you should use in comparing perplexity fairly.  

How do you model `p(actual word | clipped word class)`?  Use a uniform distribution, a unigram language model, or even call an n-gram language model (remember to apply Bayes's rule to restrict to clipped words).  Treat the task as if you had to actually generate the entire output sentence.  

With models over characters and subword units (hi @rsennrich), neural vocabulary limitations are slowly going away.  So arguably, you'd get better results by training both KN (with larger N) and neural networks over subword units.
